AK/HashTable.h

```diff
@@ -483,36 +497,55 @@ class HashTable {
             return nullptr;
 
         hash %= m_capacity;
+        size_t probe_length = 0;
         for (;;) {
-            auto* bucket = &m_buckets[hash];
-            if (bucket->state == BucketState::Free)
+            auto state = m_states[hash];
+
+            if (state == BucketState::Free)
                 return nullptr;
-            if (predicate(*bucket->slot()))
-                return bucket;
+            if (to_underlying(state) < probe_length)
+                return nullptr;
+
+            if (state == bucket_state_for_probe_length(probe_length)) {
+                auto* bucket = &m_buckets[hash];
+                if (predicate(*bucket->slot()))
+                    return bucket;
+            }
```

Comment:
FYI, I tried keeping track of the probe length during the lookup as well as part of the original implementation, and I kept seeing worse results in benchmarks. Make sure to measure this thoroughly!